src/utils/deferNetworkRequests.ts

```diff
@@ -0,0 +1,38 @@
+/**
+ * Intercepts and defers any requests on the page
+ * until the Service Worker instance is ready.
+ */
+export function deferNetworkRequests(
+  workerReady: Promise<ServiceWorkerRegistration | null>,
+) {
+  // Defer `XMLHttpRequest` until the Service Worker is ready.
+  const originalXhrSend = window.XMLHttpRequest.prototype.send
+  const restoreXhrSend = function (
+    this: XMLHttpRequest,
+    ...args: Parameters<XMLHttpRequest['send']>
+  ) {
+    window.XMLHttpRequest.prototype.send = originalXhrSend
+    this.send(...args)
+  }
```

Comment:
These `restoreX` functions both restore *and* call the function - maybe they could be named in a way where that's clearer?

Reply:
I'll adopt `performActualX`:

- `performActualXMLHttpRequest`
- `performActualFetch`

Reply:
Removed the restoration functions at all, used `until` that helped to resolve the Promise, regardless whether it resolves/rejects.